--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -5,6 +5,7 @@
 import net.minecraft.world.item.ItemStack;
 
 public abstract class CustomRecipe implements CraftingRecipe {
+
     private final ResourceLocation id;
     private final CraftingBookCategory category;
 
@@ -32,4 +33,11 @@
     public CraftingBookCategory category() {
         return this.category;
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe() {
+        return new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftComplexRecipe(this);
+    }
+    // CraftBukkit end
 }
